src/components/admin/form-components/withSearchBar.jsx

```diff
@@ -1,223 +1,94 @@
 import { cleanupFalcorKeys } from 'lib/falcor/falcor-utilities';
+import { debounce } from 'lib/utilities';
+import { getDisplayName } from 'lib/higher-order-helpers';
 
-// material-ui
-import TextField from 'material-ui/TextField';
-import Menu from 'material-ui/Menu';
-import MenuItem from 'material-ui/MenuItem';
-import ContentAdd from 'material-ui/svg-icons/content/add';
-
-const withSearchBar = config => {
-  if (!('formatter' in config) || !_.isFunction(config.formatter)) {
-    throw new Error(
-      'withSearchBar.config.formatter must be supplied as a function.',
-    );
-  }
-  if (!('fields' in config) || !_.isArray(config.fields)) {
-    throw new Error(
-      'withSearchBar.config.fields must be supplied as an array.',
-    );
-  }
-  if (!('displayName' in config) || !_.isString(config.displayName)) {
-    throw new Error(
-      'withSearchBar.config.displayName must be supplied as a string.',
-    );
-  }
-  if (!('category' in config) || !_.isString(config.category)) {
-    throw new Error(
-      'withSearchBar.config.category must be supplied as a string.',
-    );
-  }
-
-  class SearchBar extends BaseComponent {
+export const withSearchableConcept = (
+  fields,
+  formatter,
+  category,
+) => WrappedField => {
+  const debounceTime = 250;
+  class Searchable extends React.Component {
     constructor(props) {
       super(props);
-      this.fieldUpdaters = {
-        searchValue: updateFieldValue.bind(this, 'searchValue', undefined),
+      this.state = {
+        suggestions: [],
       };
-      this.safeSetState({
-        searchValue: '',
-        searchSuggestions: [],
-      });
 
-      this.debouncedGetSuggestions = debounce(() => {
-        const query = this.state.searchValue;
+      this.handleChange = this.handleChange.bind(this);
+      this.debounceSuggestions = debounce(query => {
         if (!query.trim()) {
-          this.safeSetState({ searchSuggestions: [] });
+          this.setState({ suggestions: [] });
           return;
         }
 
         let pathSets = this.props.extraPathSets || [];
         pathSets = pathSets.reduce(
           (arr, pathSet) =>
             arr.push(
-              ['search', 'posts', query, { length: this.props.length }].concat(
+              ['search', category, query, { length: this.props.length }].concat(
                 pathSet,
               ),
             ),
           [
             [
               'search',
-              config.category,
+              category,
               query,
               { length: this.props.length },
-              config.fields,
+              _.uniq([...fields, ...this.props.fields]),
             ],
           ],
         );
-
-        this.props.model.get(...pathSets).then(x => {
+        this.props.falcor.get(...pathSets).then(x => {
           if (!x) {
-            this.safeSetState({ searchSuggestions: [] });
+            this.setState({ suggestions: [] });
```

Comment:
As above, I don't think suggestions should live up here. The only reason it's up here seems to be because of this setState logic. If I understand what is going on here, I think this logic could be handled in the update suggestions function. Then, the component would set it's own "suggestions" and the only thing it should really need to pass upwards is the item that the user clicked.

Reply:
Hey Will, thanks for the comments -- appreciate them! In the general design pattern I wanted to separate the logic (fetching search results) from the view (the SearchBar). In this way, search could be used on any component that takes suggestions and not just a SearchBar. I am not sure if this is needed, but I think it provides a nice abstraction. 

If I push the search logic into the searchbar, won't I then have the model and the view together? This is what I tried to avoid by pushing update suggestions outside the searchbar component. But maybe this design is wrong? If there is no need for the updateSuggestions to be outside of searchbar, then I don't see a need for the HOC. Do you think the HOC is still justified though?

Reply:
I totally agree that the logic and the view should be separated! Really all I am saying is that rather than having the `debouncedQuery` function use `this.setState(suggestions)` in the HOC we can instead make `debouncedQuery` simply `return(suggestions)`. 

Then, inside of searchBar you can just use `this.setState(this.props.updateSuggestions)` so that `SearchBar` owns all of the state which only it displays.

Reply:
Yup, I'll implement this.